PROCESS.md

```diff
@@ -0,0 +1,171 @@
+
+### STAGE-2
+`STAGE-2` features are in the last stages of becoming stable. Changes are
+unlikely, but backward incompatible changes are still allowed in extreme cases.
+Users can be reasonably sure that the feature won't change, but can't expect all
+implementations to support the feature yet.
+
+`STAGE-2` is a staging area for features that are ready to be declared stable,
+but are waiting for the next release or don't yet have enough implementations.
+When we see the vast majority of actively maintained implementations support
+this feature, it will be promoted to stable in the next release. Implementers
+SHOULD support `STAGE-2` features so they aren't out-of-date when the next
+release happens.
+
+Being in `STAGE-2` isn't a guarantee that a feature will be declared stable in
+the next release. If we don't see that the vast majority of actively maintained
+implementation have implemented the feature, it may stay in `STAGE-2` another
+year to allow implementations to catch up.
+
+`STAGE-2` features can only be promoted to stable as part of a release.
+
+**Promotion to Stable Checklist:**
+[] The vast majority of actively maintained implementations support the feature
+
+### Stable
+Any part of the specification that doesn't have stability flag or has a year
+stability flag such as `2024` is considered stable. This flag indicates the year
+a feature was declared stable. Any changes to stable features MUST be backward
+compatible and not introduce problems for forward compatibility. Stable features
+are never removed, but may be deprecated.
+
+### Deprecated
+Any feature in the specification that is flagged in the form `DEPRECATED-{YEAR}`
+is considered deprecated as of the year specified. Implementations SHOULD still
+support these features to maintain backward compatibility. Deprecated features
+will never be removed from the spec, but schema authors SHOULD migrate away from
+using them as implementations may begin dropping support for these features at
+some point.
+
+A feature can only be deprecated as part of a release.
+
+## Release Process
+Releases occur on Jan 1 every year. Each release is identified with the
+four-digit year of when the release occurred. Releases never introduce anything
+new. The only change should be updating `STAGE-2` flags to `{YEAR}` flags for
+the features that are promoted to stable that year, or adding adding deprecation
+flags for any features that are deprecated that year.
+
+Implementations that express support for a particular release MUST support all
+features that are stable as of that release.
+
+A snapshot of the spec will be taken of the stable parts of the spec and made
+available as a PDF on the website.
+
+## Meta-Schemas
+Meta-schemas associated with the specification(s) maintained under this process
+are subject to the same compatibility rules as the specification. They can be
+updated at any time, but changes MUST be backward compatible and allow for
+forward compatibility with future schemas.
```

Comment:
Just like the spec, there is one set of mutable meta-schemas that describe the current state of the specification. Because of backward and forward compatibility constraints, no matter what release of the spec the schema is written for, the latest meta-schemas is all you need.

> Are we going to maintain meta-schemas attached to individual release snapshots?

We could have snapshots for reference like with the spec, but there wouldn't be reason to actually use those snapshots for anything other than historical reference.

> Are we going to have a "current" meta-schema which contains all keywords, even stage-0?
> Are we going to have a "current" meta-schema which per stage? For example, all stage-1 and above or all stage-2 and above?

The meta-schemas should include anything that's in the spec. That includes `STAGE-1`, `STAGE-2`, and deprecated.

> If we have a "current" meta-schema, which includes all of the stages, how can we update meta-schemas yet maintain backwards compatiblity?

Meta-schema changes should be backward compatible because spec changes are backward compatible.

> This becomes a very specific problem if we disallow unknown keywords by default.
>
> What if we wanted to update the meta-schemas to use a new keyword? That wouldn't be backwards compatible.

I was thinking that the standard dialect meta-schema would allow additional keywords. If a keyword appears in the meta-schema but is not supported by the implementation, the implementation should raise an error. It's not something the meta-schema can know, so I consider it a runtime constraint the implementation needs to deal with, not the meta-schema.

However, meta-schemas could disallow additional keywords and it shouldn't hurt anything. If an implementation is using an older version of the meta-schema that doesn't have a newer keyword used by the schema, then the implementation also wouldn't have support for the keyword. The result is always that you get an error that the implementation doesn't know the keyword.

Either way, there shouldn't be a backward or forward compatibility issue and whether we choose to allow additional keywords in the meta-schema is not relevant to process, so we can discuss that when we get to writing the meta-schemas.

Reply:
I expect there to be some discussion about meta-schema lifecycle that may end up as a bigger conversation. This was the kind of thing I thought might be best to get into after the initial merge. Like I said, I don't expect merging this document to mean it's "done". I expect it to be a first draft that we will iterate on.

Reply:
There are a number of assertions in your response which I don't agree with, but I don't think here's the right place to capture this discussion.

>I expect there to be some discussion about meta-schema lifecycle that may end up as a bigger conversation. This was the kind of thing I thought might be best to get into after the initial merge. Like I said, I don't expect merging this document to mean it's "done". I expect it to be a first draft that we will iterate on.

I think we need to discuss how we can break up the SDLC into more discreet chunks for further discussion.

Reply:
That's exactly why I proposed merging this as a work-in-progress and opening issues to discuss the parts that need more discussion. Holding this up until we decide every detail makes it more difficult to have those discussions.